R/output.R

```diff
@@ -69,6 +84,45 @@ possibly <- function(.f, otherwise, quiet = TRUE) {
   }
 }
 
+#' @export
+#' @rdname safely
+auto_browse <- function(f) {
+  if (is_primitive(f)) {
+    abort("Can not auto_browse() primitive functions")
+  }
+
+  function(...) {
+    withCallingHandlers(
+      f(...),
+      error = function(e) {
+        # 1: h(simpleError(msg, call))
+        # 2: .handleSimpleError(function (e)  <...>
+        # 3: stop(...)
+        frame <- eval_frame(4)
+        browse_in_frame(frame)
+      },
+      warning = function(e) {
+        if (!identical(getOption("warn"), 2)) {
+          frame <- eval_frame(7)
+          browse_in_frame(frame)
+        }
+      }
+    )
+  }
+}
+
+browse_in_frame <- function(frame) {
+  if (.Platform$GUI == "ESS") {
+    # Workaround ESS issue
+    with_env(frame$env, on.exit({
+      browser()
+      NULL
+    }))
+    return_from(frame)
+  } else {
+    eval(quote(browser()), envir = frame$env)
```

Comment:
this should be `expr_eval()`, since `eval()` is not frame consistent (it creates a new frame that has `envir` as frame environment, so frame-dependent functions get tricked when they look up the eval stack.

I think `expr_eval()` should have a shorter name and become our default way of evaluating code. `base::eval()` has much more complicated logic due to the possibility of adding auxiliary data, and creates a quirky call stack. Maybe `exec()`? Not sure if this feels right, "eval" is such an established term.